### PR TITLE
Savepoints

### LINKED ISSUE #4876: Use savepoints during globaldb updates
At the moment we apply all the updates at once when the conflict method resolution has been chosen. This presents the issue where if the sql query is incorrect we can end up with missing assets. The reason for this is this delete statement https://github.com/rotki/rotki/blob/develop/rotkehlchen/globaldb/updates.py#L69 that is not rolled back at https://github.com/rotki/rotki/blob/develop/rotkehlchen/globaldb/updates.py#L353 and if we try to roll back we undo all the previous actions.

For this lefteris suggested that we use https://www.sqlite.org/lang_savepoint.html in the update process

### PR DESCRIPTION
**Pull request by nebolax**
Closes #4876 

## Checklist

- [x] Introduce savepoints in gevent driver
- [x] Use savepoints during assets updates

### CODE REVIEW

File: rotkehlchen/globaldb/updates.py [hunk 1]
```diff
@@ -301,7 +301,7 @@ def _parse_full_insert(self, insert_text: str) -> AssetData:
 
     def _apply_single_version_update(
             self,
-            cursor: DBCursor,
+            connection: DBConnection,
             version: int,
             text: str,
             conflicts: Optional[Dict[Asset, Literal['remote', 'local']]],
```
Comment (prettyirrelevant) [inline rotkehlchen/globaldb/updates.py:307]: Shouldn't this entire method use a savepoint also? So we have something like...
-> SAVEPOINT x
-----> SAVEPOINT x
-----> SAVEPOINT x

In this case, if the inner save goes through but for some reason, an exception occurs in the outer savepoint, the whole thing is rolled back.
What do you think?

**Review by prettyirrelevant — COMMENTED** [review]
@nebolax left you some comments

File: rotkehlchen/tests/unit/test_globaldb_asset_updates.py [hunk 2]
```diff
@@ -295,3 +296,45 @@ def test_parse_full_insert(assets_updater, text, expected_data, error_msg):
...
+    assert connection.execute('SELECT name FROM assets WHERE identifier="ETH"').fetchone()[0] == 'Ethereum'  # noqa: E501
+    assert connection.execute('SELECT symbol FROM common_asset_details WHERE identifier="ETH"').fetchone()[0] == 'ETH'  # noqa: E501
+    # BTC should have been updated since the insertion were correct
+    assert connection.execute('SELECT name FROM assets WHERE identifier="BTC"').fetchone()[0] == 'name2'  # noqa: E501
+    assert connection.execute('SELECT symbol FROM common_asset_details WHERE identifier="BTC"').fetchone()[0] == 'symbol2'  # noqa: E501
+    # NEW-ASSET-1 should have not been added since the insertions were incorrect
+    assert connection.execute('SELECT * FROM assets WHERE identifier="NEW-ASSET-1"').fetchone() is None  # noqa: E501
+    # NEW-ASSET-2 should have been added since the insertions were correct
+    assert connection.execute('SELECT name FROM assets WHERE identifier="NEW-ASSET-2"').fetchone()[0] == 'name4'  # noqa: E501
+    assert connection.execute('SELECT symbol FROM common_asset_details WHERE identifier="NEW-ASSET-2"').fetchone()[0] == 'symbol4'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_globaldb_asset_updates.py:340]: nice but as I mentioned in the example you need specific unit tests of the save functionality you made. You need to test that `with db.savepoint_ctx()` works properly for multiple scenarios, and also offer non context manager functions to roll back to savepoints.

Generally a proper driver/api needs to exist for this in `DBConnection` and be thoroughly unit tested (in the proper test file)

**Review by LefterisJP — COMMENTED** [review]
left comments. Not sure this is correct

File: rotkehlchen/tests/unit/test_savepoints.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import sqlite3
+
+import pytest
+
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+
+
+def test_savepoints():
+    conn = GlobalDBHandler().conn
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_savepoints.py:9]: no no no don't use the global DB like this in tests. If you wanted to use it in tests just use the globaldb fixture.

But again no this is not what I meant.

Skip our global DB completely. There is no need to involve it whatsoever here.

Just do a 
```python
conn = DBConnection(path=':memory:', connection_type=DBConnectionType.GLOBAL, sql_vm_instructions_cb=0)
```

and you have a full DB connection through our driver in memory. (This is pseudocode -- I have not tested it -- may need tweaking)

File: rotkehlchen/tests/unit/test_savepoints.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    with pytest.raises(sqlite3.Error):
+        with conn.savepoint_ctx('point1') as cursor1:
+            assert conn.savepoints == ['point1']
+            cursor1.execute('INSERT INTO a VALUES (1), (2)')  # insert something
+            with conn.savepoint_ctx('point2') as cursor2:
+                assert conn.savepoints == ['point1', 'point2']
+                # insert more and check multiple nested savepoints
+                cursor2.execute('INSERT INTO a VALUES (3)')
+                with pytest.raises(sqlite3.Error):
+                    # Also check savepoints with the same name. In this case, according to sqlite's
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_savepoints.py:20]: We can and should protect against this as I said in the other part of the review

File: rotkehlchen/tests/unit/test_savepoints.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            assert conn.savepoints == ['point1', 'point4']
+            cursor4.execute('DELETE FROM a WHERE b=1')
+            conn.release_savepoint('point4')  # test manual release
+            assert conn.savepoints == ['point1']
+            # check that 1 was removed
+            assert cursor1.execute('SELECT 1 IN (SELECT b FROM a)').fetchone()[0] == 0
+            # Finally break point1 and check that all insertions/deletions were undone
+            cursor1.execute('INSERT INTO a VALUES (2)')  # 2 already exists
+
+    assert len(conn.savepoints) == 0
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_savepoints.py:43]: generally adjust the tests for what I told you to implement in the db driver part of the review. And keep 3 tests. One with named savepoints, one without giving specific names, and one with having named one at start, then go deeper with no names but skip directly back to the named one. 

Guess the last test can be combined with one of the other 2

File: rotkehlchen/tests/unit/test_savepoints.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import sqlite3
+
+import pytest
+
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+
+
+def test_savepoints():
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_savepoints.py:8]: split the test names when you implement the tests I told you. Easier to read when you got small simple unit tests, testing one thing and having explicitly and nicely named names

File: docs/changelog.rst [hunk 7]
```diff
@@ -18,6 +18,8 @@ Changelog
 * :feature:`-` An ethereum address's tokens can now be manually detected by pressing a specific button in the UI.
 * :feature:`-` Transactions involving uniswap v2 and uniswap v3 will now be properly decoded in the ethereum transactions view
 * :bug:`-` When force sync fails a proper error message is displayed
+* :bug:`-` Now if something is wrong with an asset update, it won't end up having partial information.
```
Comment (LefterisJP) [inline docs/changelog.rst:21]: where are these 2 coming from and what do they have to do with savepoints?

File: rotkehlchen/db/drivers/gevent.py [hunk 8]
```diff
@@ -214,6 +231,7 @@ def __init__(
         self.in_callback = gevent.lock.Semaphore()
         self.connection_type = connection_type
         self.sql_vm_instructions_cb = sql_vm_instructions_cb
+        self.savepoints: Dict[str, None] = {}
```
Comment (PabloCastellano) [inline rotkehlchen/db/drivers/gevent.py:234]: Maybe add a comment explaining that we use this as a way to have ordered sets?

File: rotkehlchen/db/drivers/gevent.py [hunk 9]
```diff
@@ -295,6 +313,87 @@ def write_ctx(self) -> Generator['DBCursor', None, None]:
...
+        Returns cursor and savepoint's name.
+        May raise:
+        - ContextError if a savepoint with the same name already exists. Can only happen in case of
+        manually specified name.
+        """
+        if savepoint_name is None:
+            savepoint_name = str(uuid4())
+        if savepoint_name in self.savepoints:
+            raise ContextError(
+                f'Wanted to enter savapoint {savepoint_name} but a savepoint with the same name '
```
Comment (LefterisJP) [inline rotkehlchen/db/drivers/gevent.py:349]:
```suggestion
                f'Wanted to enter savepoint {savepoint_name} but a savepoint with the same name '
```

File: rotkehlchen/db/drivers/gevent.py [hunk 10]
```diff
@@ -295,6 +313,87 @@ def write_ctx(self) -> Generator['DBCursor', None, None]:
...
+        elif savepoint_name not in self.savepoints:
+            raise ContextError(
+                f'Incorrect use of savepoints! Wanted to revert savepoint {savepoint_name}, '
+                f'but it is not present in the stack: {list_savepoints}',
+            )
+        if rollback_or_release == 'rollback':
+            self.execute(f'ROLLBACK TO SAVEPOINT "{savepoint_name}"')
+        else:  # release
+            self.execute(f'RELEASE SAVEPOINT "{savepoint_name}"')
+        self.savepoints = dict.fromkeys(list_savepoints[:list_savepoints.index(savepoint_name)])
```
Comment (PabloCastellano) [inline rotkehlchen/db/drivers/gevent.py:379]: Uhmmm, I think this will not work. Why not just `del self.savepoints[savepoint_name]`?

**Review by PabloCastellano — COMMENTED** [review]
Left a comment about a possible minor issue

File: rotkehlchen/db/drivers/gevent.py (hunk 10 above)
Comment (PabloCastellano) [reply rotkehlchen/db/drivers/gevent.py:379]: Oh nevermind, I think I understand the way it works now.

File: rotkehlchen/db/drivers/gevent.py [hunk 11]
```diff
@@ -295,6 +313,87 @@ def write_ctx(self) -> Generator['DBCursor', None, None]:
...
+        return cursor, savepoint_name
+
+    def _modify_savepoint(
+            self,
+            rollback_or_release: Literal['rollback', 'release'],
+            savepoint_name: Optional[str],
+    ) -> None:
+        if len(self.savepoints) == 0:
+            raise ContextError(
+                f'Incorrect use of savepoints! Wanted to rollback to savepoint {savepoint_name}, '
```
Comment (LefterisJP) [inline rotkehlchen/db/drivers/gevent.py:364]:
```suggestion
                f'Incorrect use of savepoints! Wanted to {rollback_or_release.lower()} savepoint {savepoint_name}, '
```

File: rotkehlchen/db/drivers/gevent.py [hunk 12]
```diff
@@ -295,6 +313,87 @@ def write_ctx(self) -> Generator['DBCursor', None, None]:
...
+            savepoint_name = list_savepoints[-1]
+        elif savepoint_name not in self.savepoints:
+            raise ContextError(
+                f'Incorrect use of savepoints! Wanted to revert savepoint {savepoint_name}, '
+                f'but it is not present in the stack: {list_savepoints}',
+            )
+        if rollback_or_release == 'rollback':
+            self.execute(f'ROLLBACK TO SAVEPOINT "{savepoint_name}"')
+        else:  # release
+            self.execute(f'RELEASE SAVEPOINT "{savepoint_name}"')
```
Comment (LefterisJP) [inline rotkehlchen/db/drivers/gevent.py:378]:
```suggestion
        self.execute(f'{rollback_or_release} SAVEPOINT "{savepoint_name}"')
```

**Review by LefterisJP — COMMENTED** [review]
nice. Some comments on improvements.

File: docs/changelog.rst (hunk 7 above)
Comment (nebolax) [reply docs/changelog.rst:21]: Partial asset info after an update is fixed in this pr.

Support of one address being on both bitcoin and bitcoin cash is coming from a fix that I made a couple of weeks ago. But realized that I didn't add a changelog entry only now.

File: rotkehlchen/db/drivers/gevent.py (hunk 10 above)
Comment (nebolax) [reply rotkehlchen/db/drivers/gevent.py:379]: Yeah, will add a comment in the code also

File: docs/changelog.rst (hunk 7 above)
Comment (LefterisJP) [reply docs/changelog.rst:21]: Isn't that already included in 1.25.3 changelog?

Comment (nebolax) [reply docs/changelog.rst:21]: Ah wow, yes indeed

**Review by LefterisJP — APPROVED** [review]
LGTM

**Pull request merged**